pkg/ui/server.go

```diff
@@ -0,0 +1,204 @@
+		Addr:         ":9091",
+		Handler:      router,
+		ReadTimeout:  5 * time.Second,
+		WriteTimeout: 1 * time.Second,
+		IdleTimeout:  15 * time.Second,
+	}
+
+	fs := http.FileServer(http.Dir("./assets/static"))
+	router.Handle("/assets/static/", http.StripPrefix("/assets/static/", fs))
+	router.HandleFunc("/popup", popup)
+	log.Fatal(server.ListenAndServe())
+}
+
+func popup(w http.ResponseWriter, r *http.Request) {
+	details := FormDetails{
+		WindowName: r.FormValue("windowname"),
+		Origin:     r.FormValue("origin"),
+		PageTitle:  "JVS - Justification Request System",
+		Category:   r.FormValue("category"),
+		Reason:     r.FormValue("reason"),
+		TTL:        r.FormValue("ttl"),
+		Content: Content{
+			UserLabel:     "User",
+			CategoryLabel: "Category",
+			ReasonLabel:   "Reason",
+			TTLLabel:      "TTL",
+			Categories: []Pair{
+				{
+					Key:  "explanation",
+					Text: "Explanation",
+				},
+				{
+					Key:  "breakglass",
+					Text: "Breakglass",
+				},
+			},
+			TTLs: []Pair{
+				{
+					Key:  "15",
+					Text: "15m",
+				},
+				{
+					Key:  "30",
+					Text: "30m",
+				},
+				{
+					Key:  "60",
+					Text: "1h",
+				},
+				{
+					Key:  "120",
+					Text: "2h",
+				},
+				{
+					Key:  "240",
+					Text: "4h",
+				},
+			},
+		},
+	}
+
+	// Initial page load, just render the page
+	if r.Method == "GET" {
+		// set some defaults
+		details.Category = categories[0]
+		details.TTL = ttls[0]
+		render(w, "./assets/templates/index.html.tmpl", details)
+		return
+	}
+
+	// Form submission
+	if r.Method == "POST" {
+		// 1. Validate input
+		if !details.Validate() {
+			render(w, "./assets/templates/index.html.tmpl", details)
+			return
+		}
+
+		// 2. [TODO] Request a token
+		token := "token_from_server"
+
+		// 3. Redirect to a confirmation page with context, ultimately needed to postMessage back to the client
+		successContext := SuccessDetails{
+			Token:        token,
+			TargetOrigin: details.Origin,
+			WindowName:   details.WindowName,
+		}
+
+		render(w, "./assets/templates/success.html.tmpl", successContext)
+	}
+}
+
+func (formDetails *FormDetails) Validate() bool {
+	formDetails.Errors = make(map[string]string)
+
+	if !isValidOneOf(formDetails.Category, categories) {
+		formDetails.Errors["Category"] = "Category must be selected"
+	}
+
+	if strings.TrimSpace(formDetails.Reason) == "" {
+		formDetails.Errors["Reason"] = "Reason is required"
+	}
+
+	if !isValidOneOf(formDetails.TTL, ttls) {
+		formDetails.Errors["TTL"] = "TTL is required"
+	}
+
+	return len(formDetails.Errors) == 0
+}
+
+func isValidOneOf(selection string, options []string) bool {
+	for _, v := range options {
+		if v == selection {
+			return true
+		}
+	}
+	return false
+}
+
+func render(w http.ResponseWriter, filename string, data interface{}) {
+	tmpl, err := template.ParseFiles(filename)
+	if err != nil {
+		log.Print(err)
+		http.Error(w, "Sorry, something went wrong", http.StatusInternalServerError)
+		return
+	}
```

Comment:
We don't want to parse files on each request. This should be done in `main` as part of boot.

Reply:
Ack